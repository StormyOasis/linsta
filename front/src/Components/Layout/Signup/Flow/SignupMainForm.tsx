import React from "react";
import { styled } from "styled-components";
import { Link } from "react-router-dom";

import FBIconSVG from "/public/images/facebook.svg";
import LogoSVG from "/public/images/linsta.svg";

import * as styles from "../SignupLayout.module.css";
import Theme from "../../../../Components/Themes/Theme";
import SignupInput from "../Common/SignupInput";
import SignupButton from "../Common/SignupButton";
import {postAccountsAttempt, getAccountsCheckUserUnique} from "../../../../api/ServiceController";

const LogoWrapper = styled.div`
  align-content: stretch;
  align-items: stretch;
  border: none;
  display: flex;
  flex-direction: column;
  flex-grow: 0;
  flex-shrink: 0;
  justify-content: flex-start;
  overflow: visible;
  position: relative;
  margin-bottom: 12px;
  margin-top: 36px;
`;

const LoginWithFBButton = styled.button<{ $props?: any }>`
  border-radius: 8px;
  text-decoration: none;
  align-items: center;
  font-weight: 600;
  justify-content: center;
  text-wrap: nowrap;
  color: ${props => props.theme['colors'].buttonTextColorDefault};
  background-color: ${(props) => props.theme["colors"].buttonDefaultColor};
  cursor: pointer;
  display: flex;
  height: 34px;
  position: relative;
  text-align: center;
  padding-left: 16px;
  padding-right: 16px;
  flex-direction: row;
  border: none;
  margin-bottom: 12px;

  &:hover {
    background-color: ${(props) => props.theme["colors"].buttonOnHoverColor};
  }
`;

const SignupForm = styled.form`
  display: flex;
  flex-direction: column;
  margin: 0;
  max-width: 350px;
  vertical-align: baseline;
`;

type MainSignupFormProps = {
    emailOrPhone: string;
    fullName: string;
    userName: string;
    password: string;
    emailOrPhone_valid: boolean;
    fullName_valid: boolean;
    userName_valid: boolean;
    password_valid: boolean;
    changePage: any;
    handleFormChange: any;
};

export default class MainSignupForm extends React.Component<MainSignupFormProps> {
    loginWithFacebookClicked = (event: React.MouseEventHandler<HTMLButtonElement>) => {
        window.alert("todo");
    };

    validateEmailPhone = (value: string): boolean => {
        if (value == null) {
            return false;
        }

        const phoneRegex = /^\(?([0-9]{3})\)?[-. ]?([0-9]{3})[-. ]?([0-9]{4})$/;
        const emailRegex =
            /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/;
        return emailRegex.test(value) || phoneRegex.test(value);
    };

    validateFullName = (value: string): boolean => {
        return (
            value !== null &&
            value.trim().length > 0 &&
            value.trim().split(" ").length > 1
        );
    };

    validatePassword = (value: string): boolean => {
        const regex =
            /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@.#$!%*?&^])[A-Za-z\d@.#$!%*?&]{8,15}$/;

        return regex.test(value);
    };

    validateUserName = async (value: string): Promise<boolean> => {
        if (value == null || value.length == 0) return false;

        const response = await getAccountsCheckUserUnique(value);

        return response.data;
    };

    validateField = (event: React.ChangeEvent<HTMLInputElement>): boolean => {
        let result = false;
        switch (event.target.name) {
            case "emailOrPhone": {
                result = this.validateEmailPhone(event.target.value);
                break;
            }
            case "fullName": {
                result = this.validateFullName(event.target.value);
                break;
            }
            case "password": {
                result = this.validatePassword(event.target.value);
                break;
            }
        }

        return result;
    };

    handleFormChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        this.props.handleFormChange(
            event.target.name,
            event.target.value,
            this.validateField(event)
        );
    };

    handleUsernameFormChange = async (
        event: React.ChangeEvent<HTMLInputElement>
    ) => {
        const value = event.target.value;
        const name = event.target.name;

        await this.validateUserName(value)
            .then((res) => {
                this.props.handleFormChange(name, value, !res);
            })
            .catch(() => {
                this.props.handleFormChange(name, value, false);
            });
    };

    isFormValid = (): boolean => {
        return (
            this.props.emailOrPhone_valid &&
            this.props.fullName_valid &&
            this.props.password_valid &&
            this.props.userName_valid
        );
    };

    submitForm = async (event: React.ChangeEvent<HTMLButtonElement>) => {
        event?.preventDefault();

        if (!this.isFormValid()) {
            return false;
        }

        const res = await postAccountsAttempt({
            emailOrPhone: this.props.emailOrPhone,
            userName: this.props.userName,
            fullName: this.props.fullName,
            password: this.props.password,
        });

        if (res.statusText === "OK") {
            this.props.changePage(1);
            return true;
        }

        return false;
    };

    override render() {
        return (
            <Theme>
                <LogoWrapper>
                    <Link to="/">
                        <LogoSVG />
                    </Link>
                </LogoWrapper>
                <div>
                    <SignupForm method="post">
                        <div className={styles.signupFormDiv1}>
                            <span className={styles.signupIntroSpan}>
                                Sign up to see photos and videos from your friends.
                            </span>
                        </div>
                        <div className={styles.signupFormFbLoginDiv}>
                            <LoginWithFBButton onClick={this.loginWithFacebookClicked}>
                                <span className={styles.signupFormFbIcon}>
                                    <FBIconSVG />
                                </span>
                                Log in with Facebook
                            </LoginWithFBButton>
                            <div className={styles.signupFormDiv1}>
                                <div className={styles.signupFormDiv2}>
                                    <div className={styles.signupFormDiv3} />
                                    <div className={styles.signupFormDiv4}>OR</div>
                                    <div className={styles.signupFormDiv3} />
                                </div>
                            </div>
                        </div>
                        <SignupInput
                            name="emailOrPhone"
                            placeholder="Phone Number or Email"
                            value={this.props.emailOrPhone}
                            isValid={this.props.emailOrPhone_valid}
                            onChange={this.handleFormChange}
                        ></SignupInput>
                        <SignupInput
                            name="fullName"
                            placeholder="Full Name"
                            value={this.props.fullName}
                            isValid={this.props.fullName_valid}
                            onChange={this.handleFormChange}
                        ></SignupInput>
                        <SignupInput
                            name="userName"
                            placeholder="Username"
                            value={this.props.userName}
                            isValid={this.props.userName_valid}
                            onChange={this.handleUsernameFormChange}
                        ></SignupInput>
                        <SignupInput
                            name="password"
                            placeholder="Password"
                            type="password"
                            value={this.props.password}
                            isValid={this.props.password_valid}
                            onChange={this.handleFormChange}
                        ></SignupInput>

                        <div className={styles.termsDiv}>
                            By signing up, you agree to our Terms, Privacy Policy and Cookies
                            Policy.
                        </div>
                        <SignupButton
                            type="button"
                            disabled={!this.isFormValid()}
                            onClick={this.submitForm}
                            text="Sign Up"
                        />
                    </SignupForm>
                </div>
            </Theme>
        );
    }
}
