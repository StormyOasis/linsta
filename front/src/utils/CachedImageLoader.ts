import { clear, del, get, set } from 'idb-keyval';
import { blobToBase64 } from "./utils";

export const loadImageCached = async (id: string, url: string): Promise<string> => {
    // First try to get the image from the cache
    const value = await get(id);

    if(value === undefined) {
        // Image not in the cache so have to fetch it
        try {
            const result = await fetch(url);
            if(result.status === 200) {
                const blob = await result.blob();
                const b64:string = await blobToBase64(blob) as string;

                // add the base64 data into the cache and return it
                await set(id, b64);

                return b64;
            }
        } catch(err) {
            console.log(err);
            return url; //default to what was passed in
        }

    } else {
        // Image is in the cache already as a base64 string, use that as is
        return value;
    }

    return url; //default to what was passed in
}

export const setImage = async (id: string, data: string) => {     
    await set(id, data);
}

export const deleteFromCache = async (id: string) => {
    // First try to get the image from the cache
    const value = await get(id);

    if(value === undefined) {
        // key not present in cache so do nothing
        return;
    }

    // delete from the cache
    await del(id);
}

export const clearCache = async () => {
    await clear();
}