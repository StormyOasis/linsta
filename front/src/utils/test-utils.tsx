import React, { ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { RootState } from "../Components/Redux/redux";
import { authInitialState } from '../Components/Redux/slices/auth.slice';
import { modalInitialState } from '../Components/Redux/slices/modals.slice';
import { profileInitialState } from '../Components/Redux/slices/profile.slice';
import { miscInitialState } from '../Components/Redux/slices/misc.slice';
import { configureStore } from '@reduxjs/toolkit';

import Theme from '../Components/Themes/Theme';

export const mockInitialState = (overrides?: Partial<RootState>): RootState => ({
    auth: { ...authInitialState, ...(overrides?.auth ?? {}) },
    modal: { ...modalInitialState, ...(overrides?.modal ?? {}) },
    profile: { ...profileInitialState, ...(overrides?.profile ?? {}) },
    misc: { ...miscInitialState, ...(overrides?.misc ?? {}) },
});

type ExtendedRenderOptions = {
    store?: ReturnType<typeof configureStore>;
    preloadedState?: any;
    route?: string; // ‚Üê allow overriding the initial route
} & Omit<RenderOptions, "queries">;

const dummyAuthReducer = (state = { user: null, isLoading: false, error: null }, _action: any) => state;
const dummyModalReducer = (state = { isOpen: false }, _action: any) => state;
const dummyProfileReducer = (state = { name: '', email: '' }, _action: any) => state;
const dummyMiscReducer = (state = { someData: null }, _action: any) => state;

const renderWithStore = (
    ui: ReactElement,
    {
        preloadedState,
        store = configureStore({
            reducer: {
                auth: dummyAuthReducer,
                modal: dummyModalReducer,
                profile: dummyProfileReducer,
                misc: dummyMiscReducer,
            },
            preloadedState,
        }),
        route = "/",
        ...renderOptions
    }: ExtendedRenderOptions = {}
) => {
    const Wrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => {
        return (
            <Provider store={store}>
                <MemoryRouter initialEntries={[route]}>
                    <Theme>{children}</Theme>
                </MemoryRouter>
            </Provider>
        );
    };

    return {
        ...render(ui, { wrapper: Wrapper, ...renderOptions }),
        store,
    };
};

export * from "@testing-library/react";
export { renderWithStore };
export type { ExtendedRenderOptions };