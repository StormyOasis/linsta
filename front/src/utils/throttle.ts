import { useRef } from 'react';

const useThrottle = (callback: Function, delay: number) => {
    const lastCall = useRef(0);

    return (...args: any[]) => {
        const now = new Date().getTime();
        if (now - lastCall.current >= delay) {
            lastCall.current = now;
            callback(...args);
        }
    };
};

export default useThrottle;