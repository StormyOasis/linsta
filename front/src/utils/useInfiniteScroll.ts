import React, { useEffect, useRef } from 'react';
import useThrottle from './useThrottle';

/**
 * Type for the async load more function.
 */
type AsyncLoadMoreFunction = () => Promise<void>;

/**
 * Custom hook for infinite scrolling.
 * Attaches a throttled scroll event listener to a container and triggers `loadMore`
 * when the user scrolls near the bottom.
 *
 * @param loadMore - Async function to load more items.
 * @param isLoading - Whether a load is currently in progress.
 * @param containerRef - Ref to the scrollable container element.
 */
const useInfiniteScroll = (loadMore: AsyncLoadMoreFunction, 
    isLoading: boolean, 
    childRef: React.MutableRefObject<HTMLDivElement | null>) => {
        
    const hasScrolled = useRef<boolean>(false);  // Track if the user has scrolled

    // Throttled scroll handler
    const throttledHandleScroll = useThrottle(async () => {
        if (typeof window !== 'undefined' && childRef.current) {
            const element = childRef.current as HTMLElement;
            const currentScroll = window.innerHeight + element.scrollTop;

            if (currentScroll + 256 >= element.scrollHeight && !isLoading && !hasScrolled.current) {
                hasScrolled.current = true;  // Mark as scrolled at least once
                await loadMore();
            }
        }
    }, 200);

    useEffect(() => {
        const childElement = childRef.current;
        if (childElement) {
            childElement.addEventListener('scroll', throttledHandleScroll);
        }

        // Clean up event listener on unmount
        return () => {
            if (childElement) {
                childElement.removeEventListener('scroll', throttledHandleScroll);
            }
        };
    }, [throttledHandleScroll, childRef]);

    // Reset trigger flag when loading finishes
    useEffect(() => {
        if (!isLoading) {
            hasScrolled.current = false;
        }
    }, [isLoading]);
}

export default useInfiniteScroll;