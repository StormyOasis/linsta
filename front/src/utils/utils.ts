import sanitizeHtml from 'sanitize-html';

import { HistoryType, Post, Profile } from "../api/types";
import { postSetFollowStatus, postUpdatePost } from '../api/ServiceController';
import { DEFAULT_PFP } from '../api/config';

/**
 * Utility object for managing navigation and history state.
 */
export const historyUtils: HistoryType = {
    navigate: null,
    location: null,
    isServer: true
}

// Regular expressions for validation
const validatePasswordRegex: RegExp = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@.#$!%*?&^_])[A-Za-z\d@.#$!%*?&_]{8,32}$/;
const phoneRegex: RegExp = /(?:([+]\d{1,4})[-.\s]?)?(?:[(](\d{1,3})[)][-.\s]?)?(\d{1,4})[-.\s]?(\d{1,4})[-.\s]?(\d{1,9})/g;
const emailRegex: RegExp = /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/;

/**
 * Checks if a string is a hashtag (starts with '#').
 */
export const isHashtag = (text: string) => text.startsWith('#');

/**
 * Checks if a string is a mention (starts with '@').
 */
export const isMention = (text: string) => text.startsWith('@');

export const validatePassword = (value: string): boolean => {
    return validatePasswordRegex.test(value);
};

export const validateEmailPhone = (value: string): boolean => {
    if (value == null) {
        return false;
    }

    return emailRegex.test(value) || phoneRegex.test(value);
};

/**
 * Checks if a full name contains at least two words.
 */
export const validateFullName = (value: string): boolean => {
    return value?.trim().split(" ").length > 1;
};

/**
 * Validates a URL string. Empty or undefined is considered valid.
 */
export const validateUrl = (text: string | undefined): boolean => {
    if (text == null || text.length === 0) {
        return true; //No url entered should be considered valid
    }

    try {
        new URL(text);
        return true;
    } catch (e) {
        return false;
    }
}

// Supported video file extensions
const validVideoExtensions: string[] = ["avif", "ogm", "wmv", "mpg", "webm", "ogv", "mov", "asx", "mpeg", "mp4", "m4v", "avi"];
/**
 * Checks if a file path has a valid video extension.
 * @throws Error if the path is invalid.
 */
export const isVideoFileFromPath = (path: string): boolean => {
    if (!path || path.trim().length < 4) {
        throw new Error("Invalid path");
    }

    const ext = path.toLowerCase().substring(path.lastIndexOf(".") + 1);

    return validVideoExtensions.includes(ext);
}

/**
 * Checks if a MIME type string indicates a video.
 * @throws Error if the type is invalid.
 */
export const isVideoFileFromType = (type: string): boolean => {
    if (type == null || type.trim().length < 4) {
        throw new Error("Invalid type");
    }

    return type.toLowerCase().includes("video");
}

/**
 * Converts a base64 string to a File blob object.
 */
export const base64ToBlob = (base64String: string, outFileName: string): File | null => {
    if (base64String === null || base64String.length === 0) {
        return null;
    }

    const contentType = base64String.substring(5, base64String.indexOf(';'));
    const data = base64String.substring(base64String.indexOf(",") + 1);

    return base64ToBlobEx(data, contentType, outFileName);
}

/**
 * Converts a base64 string (without prefix) to a File object.
 */
export const base64ToBlobEx = (base64String: string, contentType: string, outFileName: string): File => {
    const sliceSize = 512;
    const byteCharacters = atob(base64String);
    const byteArrays = [];

    for (let offset = 0; offset < byteCharacters.length; offset += sliceSize) {
        const slice = byteCharacters.slice(offset, offset + sliceSize);
        const byteNumbers = new Array(slice.length);

        for (let i = 0; i < slice.length; i++) {
            byteNumbers[i] = slice.charCodeAt(i);
        }

        byteArrays.push(new Uint8Array(byteNumbers));
    }

    return new File(byteArrays, outFileName, { type: contentType });
}

/**
 * Converts a Blob to a base64 string.
 */
export const blobToBase64 = async (blob: any): Promise<string | ArrayBuffer | null> => {
    blob = await createBlob(blob);

    return new Promise((resolve, _reject) => {
        const reader = new FileReader();
        reader.onloadend = () => resolve(reader.result);
        reader.readAsDataURL(blob);
    });
}

/**
 * Fetches a resource as a Blob.
 */
export const createBlob = async (url: string): Promise<Blob | null> => {
    try {
        const response = await fetch(url);
        const data = await response.blob();
        return data;
    } catch (err) {
        console.log(err);
    }
    return null;
}

/**
 * Extracts the MIME type from a base64 data URL.
 */
export const extractMimeTypeFromBase64 = (data: string): string | null => {
    if (data == null || data.length === 0) {
        return null;
    }

    return data.substring(5, data.indexOf(';') + 1);
}

/**
 * Extracts a frame from a video element as a JPEG blob URL.
 */
export const extractFrameFromVideo = async (video: HTMLVideoElement): Promise<string | null> => {
    if (video == null) {
        return null;
    }

    const canvas = document.createElement("canvas");
    const ctx = canvas.getContext('2d');

    if (ctx == null) {
        return null;
    }

    canvas.height = video.videoHeight || parseInt(video.style.height);
    canvas.width = video.videoWidth || parseInt(video.style.width);

    ctx.drawImage(video, 0, 0);

    return new Promise((resolve, _reject) => {
        canvas.toBlob((file: any) =>
            resolve(URL.createObjectURL(file))
            , 'image/jpeg')
    })
}

/**
 * Returns a human-readable difference between now and the given date.
 */
export const dateDiff = (dateTime: Date): string => {
    if (dateTime == null) {
        return "now";
    }

    const now = new Date().getTime();
    const dateTimeMilli = new Date(dateTime).getTime();
    const diff = new Date(now - dateTimeMilli);
    let diffMilli = diff.getTime();

    const days = Math.floor(diffMilli / 1000 / 60 / 60 / 24);
    diffMilli = diffMilli - (days * 1000 * 60 * 60 * 24);

    const weeks = Math.floor(days / 7.0);
    const hours = Math.floor(diffMilli / 1000 / 60 / 60);
    diffMilli = diffMilli - (hours * 1000 * 60 * 60);
    const mins = Math.floor(diffMilli / 1000 / 60);

    if (weeks >= 1) {
        return weeks + "w";
    }

    if (days >= 1) {
        return days + "d";
    }

    if (hours >= 1) {
        return hours + "h";
    }

    if (mins >= 1) {
        return mins + "m";
    }

    return "now";
}

/**
 * Formats a Date object as a readable string.
 */
export const getDateAsText = (date: Date): string => {
    return new Date(date).toLocaleDateString('en-us', { year: "numeric", month: "long", day: "numeric" });
}

/**
 * Finds a post by ID from a list of posts.
 */
export const getPostFromListById = (postId: string, posts: Post[]): Post => {
    return posts.find(post => post.postId === postId) ?? ({} as any);
}

/**
 * Checks if a post is liked by a given user.
 */
export const isPostLiked = (userName: string, post: Post): boolean => {
    return !!(post?.global?.likes?.some((like: any) => like.userName === userName));
}

/**
 * Toggles the liked state of a post for a user (Only on the front end, no backend call here).
 */
export const togglePostLikedState = (userName: string, userId: string, post: Post): Post => {
    if (post == null || userName == null || userId == null) {
        return post;
    }

    if (post.global.likes == null) {
        post.global.likes = [];
    }

    const index = post.global.likes.findIndex((value: any) => value.userName === userName);
    if (index === -1) {
        // Username is not in the post's like list, so add it
        post.global.likes.push({ userName, userId });
    } else {
        // Remove the username from the post's like list
        post.global.likes.splice(index, 1);
    }

    return post;
}

/**
 * Sanitizes text for safe HTML display and also returns a plain text version.
 */
export const getSanitizedText = (text: string): [string, string] => {
    if (text == null || text.length === 0)
        return [text, text];

    const textHtml = sanitizeHtml(text, {
        allowedTags: ['b', 'i', 'em', 'strong', 'a', 'br', 'sub', 'sup'],
        allowedAttributes: {
            'a': ['href']
        },
        allowedIframeHostnames: []
    }).trim();

    const textStripped = sanitizeHtml(text, {
        allowedTags: [],
        allowedAttributes: {},
        allowedIframeHostnames: []
    }).trim();

    return [textHtml, textStripped];
}

/**
 * Checks if an element with the given ID is vertically overflowed.
 */
export const isOverflowed = (id: string): boolean => {
    const element = document.getElementById(id);

    if (element == null) {
        return false;
    }

    return element.offsetHeight < element.scrollHeight;
}

/**
 * Follows or unfollows a user by calling the backend service.
 */
export const followUser = async (userId: string, followUserId: string, shouldFollow: boolean): Promise<boolean> => {
    if (userId === followUserId) {
        return true;
    }

    const data = {
        follow: shouldFollow,
        userId: userId,
        followId: followUserId
    };

    const result = await postSetFollowStatus(data);

    return result.status === 200;
}

/**
 * Returns the profile picture URL from a Profile object, or a default if missing.
 */
export const getPfpFromProfile = (profile: Profile | null): string => {
    return profile?.pfp?.length ? profile.pfp : DEFAULT_PFP;
}

/**
 * Returns the profile picture URL from a Post object, or a default if missing.
 */
export const getPfpFromPost = (post: Post): string => {
    if (post.user.pfp == null || post.user.pfp.length === 0) {
        return DEFAULT_PFP;
    }

    return post.user.pfp;
}

/**
 * Splits a full name into first, middle, and last names.
 */
export const splitFullName = (fullName: string): { firstName: string, middleNames: string, lastName: string } => {
    const names = fullName.trim().split(' ');
    const firstName = names[0];
    const lastName = names[names.length - 1];
    const middleNames = names.slice(1, names.length - 1).join(' ');

    return { firstName, middleNames, lastName };
}

/**
 * Updates fields on a post and calls onClose with the result.
 */
export const updatePostFields = async (post: Post, fieldsToUpdate: { key: string, value: any }[], onClose: (data: any) => void) => {
    if (fieldsToUpdate == null || fieldsToUpdate.length === 0) {
        return;
    }

    try {
        const newPost: Post = await updatePost(post, fieldsToUpdate);
        // Close the modal
        onClose({ isCommited: true, post: newPost, isDeleted: false });
    } catch (err) {
        console.error(err)
    }
}


/**
 * Updates a post with the given fields and returns the updated post.
 */
export const updatePost = async (post: Post, fieldsToUpdate: { key: string, value: any }[]): Promise<Post> => {
    if (fieldsToUpdate == null || fieldsToUpdate.length === 0) {
        return post;
    }

    try {
        // build the query from the given fieldsToUpdate array
        let fields: any = null;
        if (fieldsToUpdate.length > 0) {
            fields = {};
            fieldsToUpdate.forEach((field) => {
                fields[field.key] = field.value;
            });
        }

        // Call the update post service
        const results = await postUpdatePost({ postId: post.postId, fields });
        if (results == null || results.status !== 200) {
            throw new Error("Error updating post");
        }

        // update the local returned post adding the fields not returned by ES
        results.data.postId = post.postId
        results.data.user.pfp = post.user.pfp;
        results.data.commentCount = post.global.commentCount;
        if (post?.global?.likes != null) {
            results.data.global.likes = [...post?.global?.likes];
        }

        return results.data as Post;
    } catch (err) {
        console.error(err)
    }
    return post;
}

/**
 * Stores a search query (string or Profile) in localStorage under "recentSearches".
 * Prevents duplicates based on string value or Profile.profileId.
 * @param value - The search query to store.
 * @returns The updated array of stored queries.
 */
export const storeSearchQueries = (value: string | Profile): (string | Profile)[] => {
    const key: string = "recentSearches";
    try {
        const storedData = localStorage.getItem(key);
        const parsedData = storedData ? JSON.parse(storedData) : [];

        // Check if value is a duplicate
        const isDuplicate: boolean = parsedData.some((item: string | Profile) => {
            if (typeof item === 'string') {
                return item === value;
            } else {
                return item.profileId === (value as Profile).profileId;
            }
        })

        if (isDuplicate) {
            return parsedData;
        }

        parsedData.push(value);

        localStorage.setItem(key, JSON.stringify(parsedData));
        return parsedData;
    } catch (err) {
        console.error(err);
    }
    return [];
}

/**
 * Removes a search query (string or Profile) from localStorage "recentSearches".
 * @param value - The search query to remove.
 * @returns The updated array of stored queries.
 */
export const removeStoredSearchQuery = (value: string | Profile): (string | Profile)[] => {
    const key: string = "recentSearches";
    let queries: (string | Profile)[] = [];
    try {
        queries = getStoredSearchQueries();
        const filteredQueries = queries.filter((item: string | Profile) => {
            if (typeof item === 'string') {
                return item !== value;
            } else {
                return item.profileId !== (value as Profile).profileId;
            }
        });
        localStorage.setItem(key, JSON.stringify(filteredQueries));
        return filteredQueries;
    } catch (err) {
        return queries;
    }
}

/**
 * Retrieves all stored search queries (strings or Profiles) from localStorage.
 * @returns An array of stored queries.
 */
export const getStoredSearchQueries = (): (string | Profile)[] => {
    const key: string = "recentSearches";
    const item = localStorage.getItem(key);

    if (!item) {
        return [];
    }

    try {
        return JSON.parse(item);
    } catch (err) {
        return [];
    }
}

/**
 * Builds a space-separated string of user names from selected Profile objects.
 * @param selectedProfiles - A record of Profile objects keyed by string.
 * @returns A string of user names separated by spaces.
 */
export const buildCollabSearchText = (selectedProfiles: Record<string, Profile>): string => {
    if (!selectedProfiles) {
        return "";
    }

    // Clear out the text input if there's no results
    let userNames: string = "";

    Object.values(selectedProfiles).forEach((profile: Profile) => {
        userNames = `${userNames}${profile.userName} `;
    });

    return userNames.trim();
}