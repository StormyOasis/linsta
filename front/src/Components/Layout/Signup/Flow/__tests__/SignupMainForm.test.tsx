import React from "react";
import { screen, fireEvent, waitFor } from "@testing-library/react";
import SignupMainForm, { SignupMainFormProps } from "../SignupMainForm";
import { renderWithStore } from "../../../../../utils/test-utils";
import '@testing-library/jest-dom';

// Mocks
jest.mock("react-redux", () => {
  const actual = jest.requireActual("react-redux");
  return {
    ...actual,
    useDispatch: () => jest.fn(),
  };
});
jest.mock("../../../../Common/CombinedStyling", () => ({
  __esModule: true,
  Div: (props: any) => <div data-testid="mock-div" {...props} />,
  Span: (props: any) => <span data-testid="mock-span" {...props} />,
  default: {
    signupFormDiv1: "signupFormDiv1",
    signupIntroSpan: "signupIntroSpan",
    termsDiv: "termsDiv",
  },
}));
jest.mock("../../../../Themes/Theme", () => ({
  __esModule: true,
  default: ({ children }: any) => <div data-testid="mock-theme">{children}</div>
}));
jest.mock("../../../../Common/StyledInput", () => ({
  __esModule: true,
  default: (props: any) => <input data-testid={`mock-input-${props.name}`} {...props} />
}));
jest.mock("../../../../Common/StyledButton", () => ({
  __esModule: true,
  default: (props: any) => (
    <button data-testid={props.datatestid} disabled={props.disabled} onClick={props.onClick}>{props.text}</button>
  )
}));
jest.mock("../../../../Common/LargeLogo", () => ({
  __esModule: true,
  default: () => <div data-testid="mock-logo" />
}));
jest.mock("../../../../../api/ServiceController", () => ({
  postAccountsAttempt: jest.fn(() => Promise.resolve({ statusText: "OK" })),
  getAccountsCheckUserUnique: jest.fn(() => Promise.resolve({ data: false })),
}));
jest.mock("../../../../../utils/utils", () => ({
  validateEmailPhone: jest.fn(() => true),
  validateFullName: jest.fn(() => true),
  validatePassword: jest.fn(() => true),
}));

const defaultProps: SignupMainFormProps = {
  emailOrPhone: "test@example.com",
  fullName: "Test User",
  userName: "testuser",
  password: "password123",
  emailOrPhone_valid: true,
  fullName_valid: true,
  userName_valid: true,
  password_valid: true,
  changePage: jest.fn(),
  handleFormChange: jest.fn(),
};

describe("SignupMainForm", () => {
  it("renders all fields and matches snapshot", () => {
    const r = renderWithStore(<SignupMainForm {...defaultProps} />);
    expect(screen.getByTestId("mock-logo")).toBeInTheDocument();
    expect(screen.getByTestId("mock-input-emailOrPhone")).toBeInTheDocument();
    expect(screen.getByTestId("mock-input-fullName")).toBeInTheDocument();
    expect(screen.getByTestId("mock-input-userName")).toBeInTheDocument();
    expect(screen.getByTestId("mock-input-password")).toBeInTheDocument();
    expect(screen.getByTestId("submit-signupmain")).toBeInTheDocument();
    expect(r.container).toMatchSnapshot();
  });

  it("disables Sign Up button if form is invalid", () => {
    const r = renderWithStore(
      <SignupMainForm {...defaultProps} emailOrPhone_valid={false} />
    );
    expect(screen.getByTestId("submit-signupmain")).toBeDisabled();
    expect(r.container).toMatchSnapshot();
  });

  it("calls handleFormChange on input change", () => {
    const handleFormChange = jest.fn();
    renderWithStore(<SignupMainForm {...defaultProps} handleFormChange={handleFormChange} />);
    fireEvent.change(screen.getByTestId("mock-input-fullName"), { target: { name: "fullName", value: "New Name" } });
    expect(handleFormChange).toHaveBeenCalled();
  });

  it("calls handleFormChange and debounce on username change", async () => {
    const handleFormChange = jest.fn();
    renderWithStore(<SignupMainForm {...defaultProps} handleFormChange={handleFormChange} />);
    fireEvent.change(screen.getByTestId("mock-input-userName"), { target: { name: "userName", value: "newuser" } });
    // Wait for debounce
    await waitFor(() => {
      expect(handleFormChange).toHaveBeenCalled();
    });
  });

  it("calls changePage(1) on successful submit", async () => {
    const changePage = jest.fn();
    renderWithStore(<SignupMainForm {...defaultProps} changePage={changePage} />);
    fireEvent.click(screen.getByTestId("submit-signupmain"));
    await waitFor(() => {
      expect(changePage).toHaveBeenCalledWith(1);
    });
  });
});