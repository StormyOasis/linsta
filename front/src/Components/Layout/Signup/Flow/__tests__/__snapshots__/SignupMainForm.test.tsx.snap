// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignupMainForm disables Sign Up button if form is invalid 1`] = `
<div>
  <div
    data-testid="mock-theme"
  >
    <div
      data-testid="mock-theme"
    >
      <div
        data-testid="mock-logo"
      />
      <div
        data-testid="mock-div"
      >
        <form
          class="SignupMainForm__SignupForm-sc-18p437d-0 kjyUMW"
          method="post"
        >
          <div
            class="signupFormDiv1"
            data-testid="mock-div"
          >
            <span
              class="signupIntroSpan"
              data-testid="mock-span"
            >
              Sign up to see photos and videos from your friends.
            </span>
          </div>
          <input
            data-testid="mock-input-emailOrPhone"
            name="emailOrPhone"
            placeholder="Phone Number or Email"
            value="test@example.com"
          />
          <input
            data-testid="mock-input-fullName"
            name="fullName"
            placeholder="Full Name"
            value="Test User"
          />
          <input
            data-testid="mock-input-userName"
            name="userName"
            placeholder="Username"
            value="testuser"
          />
          <input
            data-testid="mock-input-password"
            name="password"
            placeholder="Password"
            type="password"
            value="password123"
          />
          <div
            class="termsDiv"
            data-testid="mock-div"
          >
            By signing up, you agree to our Terms, Privacy Policy and Cookies Policy.
          </div>
          <button
            data-testid="submit-signupmain"
            disabled=""
          >
            Sign Up
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;

exports[`SignupMainForm renders all fields and matches snapshot 1`] = `
<div>
  <div
    data-testid="mock-theme"
  >
    <div
      data-testid="mock-theme"
    >
      <div
        data-testid="mock-logo"
      />
      <div
        data-testid="mock-div"
      >
        <form
          class="SignupMainForm__SignupForm-sc-18p437d-0 kjyUMW"
          method="post"
        >
          <div
            class="signupFormDiv1"
            data-testid="mock-div"
          >
            <span
              class="signupIntroSpan"
              data-testid="mock-span"
            >
              Sign up to see photos and videos from your friends.
            </span>
          </div>
          <input
            data-testid="mock-input-emailOrPhone"
            name="emailOrPhone"
            placeholder="Phone Number or Email"
            value="test@example.com"
          />
          <input
            data-testid="mock-input-fullName"
            name="fullName"
            placeholder="Full Name"
            value="Test User"
          />
          <input
            data-testid="mock-input-userName"
            name="userName"
            placeholder="Username"
            value="testuser"
          />
          <input
            data-testid="mock-input-password"
            name="password"
            placeholder="Password"
            type="password"
            value="password123"
          />
          <div
            class="termsDiv"
            data-testid="mock-div"
          >
            By signing up, you agree to our Terms, Privacy Policy and Cookies Policy.
          </div>
          <button
            data-testid="submit-signupmain"
          >
            Sign Up
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;
